packages/opencensus-exporter-instana/src/instana.ts

```diff
@@ -0,0 +1,152 @@
+/**
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {classes, logger, types} from '@opencensus/opencensus-core';
+import {request} from 'http';
+
+type InstanaSpan = {
+  spanId: string,
+  parentId: string,
+  traceId: string,
+  timestamp: number,
+  duration: number,
+  name: string,
+  type: string,
+  error: boolean,
+  data: {[s: string]: string;}
+};
+
+export interface InstanaExporterOptions extends types.ExporterConfig {
+  agentHost?: string;
+  agentPort?: number;
+  transmissionTimeout?: number;
+}
+
+export class InstanaTraceExporter implements types.Exporter {
+  agentHost: string;
+  agentPort: number;
+  transmissionTimeout: number;
+
+  exporterBuffer: classes.ExporterBuffer;
+  logger: types.Logger;
+
+  constructor(options: InstanaExporterOptions = {}) {
+    this.agentHost =
+        options.agentHost || process.env.INSTANA_AGENT_HOST || '127.0.0.1';
+    this.agentPort =
+        options.agentPort || Number(process.env.INSTANA_AGENT_PORT) || 42699;
+    this.transmissionTimeout = options.transmissionTimeout || 10000;
+    this.logger = options.logger || logger.logger();
+    this.exporterBuffer = new classes.ExporterBuffer(this, options);
+  }
+
+  onStartSpan(root: types.RootSpan) {}
+
+  onEndSpan(root: types.RootSpan) {
+    this.exporterBuffer.addToBuffer(root);
+  }
+
+  publish(rootSpans: types.RootSpan[]): Promise<void> {
+    try {
+      return this.transmit(this.translateRootSpans(rootSpans));
+    } catch (e) {
+      this.logger.error(
+          'Unexpected error in Instana exporter during publish attempt', e);
+      return Promise.reject(e);
+    }
+  }
+
+  private translateRootSpans(rootSpans: types.RootSpan[]): InstanaSpan[] {
+    return rootSpans.reduce((agg, rootSpan) => {
+      agg.push(this.translateSpan(rootSpan));
+      return agg.concat(rootSpan.spans.map(this.translateSpan));
+    }, []);
+  }
+
+  private translateSpan(span: types.Span): InstanaSpan {
+    return {
+      spanId: span.id,
+      // Do not report parentId as null. Instead, drop the field.
+      parentId: span.parentSpanId ? span.parentSpanId : undefined,
+      traceId: span.traceId.substring(0, 8),
+      timestamp: span.startTime.getTime(),
+      duration: span.duration | 0,
```

Comment:
Should this be `|| 0`? Given that `span.duration` is already a number.

Reply:
True, `span.duration` is a number, but it might contain decimal places. This drops the decimal places. The API on Instana's side does not allow decimal places.

This is one of these things that we need to look into on Instana's end once OpenCensus gains traction. At that point, we might end up adding a different API endpoint specifically for OpenCensus. Currently, this is our generic API endpoint (which has some limitations).

Reply:
This seems to be deserving of a comment to be left in the code.

Reply:
Agreed @ofrobots and done! :)